src/api/macros/chrysalis-macros.js

```diff
@@ -249,15 +218,64 @@ class Macros {
       }
     }
     return macros;
-  }
+  };
 
-  async focus(s, macros) {
+  const expandMacro = (macro) => {
+    let m = macro.map((v) => Object.assign({}, v));
+    const mapping = {
+      KEYCODEUP: "KEYUP",
+      KEYCODEDOWN: "KEYDOWN",
+      TAPCODE: "TAP",
+      TAPCODESEQUENCE: "TAPSEQUENCE",
+    };
+
+    for (let step of m) {
+      if (mapping[step.type]) {
+        step.type = mapping[step.type];
+      }
+    }
+
+    return m;
+  };
+
+  const clone = (macro) => {
+    return macro.map((v) => Object.assign({}, v));
+  };
+
+  const compressMacro = (macro) => {
+    let m = clone(macro);
+    for (let step of m) {
+      if (step.type == "KEYUP" && step.value.code < 256)
```

Comment:
step.value.code < 256 feels like it's something that could be an "isFoo" method on step.value